Source/Core/Core/IOS/FS/HostBackend/FS.cpp

```diff
@@ -159,179 +335,307 @@ void HostFileSystem::DoState(PointerWrap& p)
+  if (!File::Rename(host_old_path, host_new_path))
   {
-    ERROR_LOG(IOS_FS, "Rename %s to %s - failed", old_name.c_str(), new_name.c_str());
+    ERROR_LOG(IOS_FS, "Rename %s to %s - failed", host_old_path.c_str(), host_new_path.c_str());
     return ResultCode::NotFound;
   }
 
+  // Finally, remove the child from the old parent and move it to the new parent.
+  const auto it = std::find_if(old_parent->children.begin(), old_parent->children.end(),
+                               GetNamePredicate(split_old_path.file_name));
+  FstEntry* new_entry = GetFstEntryForPath(new_path);
+  if (it != old_parent->children.end())
+  {
+    *new_entry = *it;
+    old_parent->children.erase(it);
+  }
+  new_entry->name = split_new_path.file_name;
+  SaveFst();
+
   return ResultCode::Success;
 }
 
-Result<std::vector<std::string>> HostFileSystem::ReadDirectory(Uid, Gid, const std::string& path)
+Result<std::vector<std::string>> HostFileSystem::ReadDirectory(Uid uid, Gid gid,
+                                                               const std::string& path)
 {
-  if (!IsValidWiiPath(path))
+  if (!IsValidPath(path))
     return ResultCode::Invalid;
 
-  // the Wii uses this function to define the type (dir or file)
-  const std::string dir_name(BuildFilename(path));
-
-  const File::FileInfo file_info(dir_name);
-
-  if (!file_info.Exists())
-  {
-    WARN_LOG(IOS_FS, "Search not found: %s", dir_name.c_str());
+  const FstEntry* entry = GetFstEntryForPath(path);
+  if (!entry)
     return ResultCode::NotFound;
-  }
 
-  if (!file_info.IsDirectory())
-  {
-    // It's not a directory, so error.
-    return ResultCode::Invalid;
-  }
+  if (!entry->CheckPermission(uid, gid, Mode::Read))
+    return ResultCode::AccessDenied;
 
-  File::FSTEntry entry = File::ScanDirectoryTree(dir_name, false);
+  if (entry->data.is_file)
+    return ResultCode::Invalid;
 
-  for (File::FSTEntry& child : entry.children)
+  const std::string host_path = BuildFilename(path);
+  File::FSTEntry host_entry = File::ScanDirectoryTree(host_path, false);
+  for (File::FSTEntry& child : host_entry.children)
   {
     // Decode escaped invalid file system characters so that games (such as
     // Harry Potter and the Half-Blood Prince) can find what they expect.
     child.virtualName = Common::UnescapeFileName(child.virtualName);
   }
 
-  // NOTE(leoetlino): this is absolutely wrong, but there is no way to fix this properly
-  // if we use the host filesystem.
-  std::sort(entry.children.begin(), entry.children.end(),
-            [](const File::FSTEntry& one, const File::FSTEntry& two) {
-              return one.virtualName < two.virtualName;
+  // Sort files according to their order in the FST tree (issue 10234).
+  // The result should look like this:
+  //     [FilesNotInFst, ..., OldestFileInFst, ..., NewestFileInFst]
+  std::unordered_map<std::string_view, int> sort_keys;
+  for (size_t i = 0; i < entry->children.size(); ++i)
+    sort_keys[entry->children[i].name] = int(i);
```

Comment:
```suggestion
    sort_keys.emplace(entry->children[i].name, int(i));
```

(or `insert_or_assign()`, whichever is preferable)